distribution/tools/server-cli/src/main/java/org/elasticsearch/server/cli/ServerCli.java

```diff
@@ -107,6 +117,9 @@ public void execute(Terminal terminal, OptionSet options, Environment env, Proce
 
         // we are running in the foreground, so wait for the server to exit
         int exitCode = server.waitFor();
+        for (final var postExitProcessor : serverProcessPostExitProcessors) {
+            postExitProcessor.process(exitCode);
```

Comment:
I think we can do this without any additional interfaces. We don't have a need for multiple, just to hook into the flow.

I suggest adding a method here, something like "checkExit", where the below if/throw can be moved. The subclass can then override, do whatever it needs, then call the super method.

Reply:
Yes, definitely. 
As you may have guessed at this point, I strongly prefer flat hierarchies and interfaces over protected methods, but I can adapt my style. It makes a lot of sense to be consistent with what it's already there. I will submit another revision in this direction